all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,340 @@
+# External Client Configuration
+
+Notes for this proposal:
+
+* Everything is subject to change both during proposal time and afterwards during implementation.
+* For each place a decision is made, a "💭 Why X? ..." aside is present explaining why.
+* For each place an open question exists, a "❓ X?" exists to ask it.
+
+## Overview
+
+There are many ways to access Temporal programmatically including CLI, SDK, and other components that leverage those
+such as samples and Terraform providers.
+
+Temporal needs a way for users to provide client configuration without hardcoding in CLI arguments or SDK options. This
+will allow code to seamlessly switch between profiles such as between a dev-server and cloud. The same configuration
+must work across all SDKs/CLIs.
+
+### Goals
+
+* Common, well-specified configuration format usable by Temporal tooling for client options
+* Implementation in all Temporal SDKs for reading and writing of this configuration
+* All Temporal CLI/SDKs updated to have easy client creation using this configuration
+* Common on-disk location in every platform for reading/writing this format
+* Ability to provide environment variables to represent values in the configuration
+* Well-specified hierarchy for on-disk overridable by environment variables overridable by in-CLI/SDK values
+* 0 new dependencies in SDKs
+* Profile names within the configuration
+
+### Non-goals
+
+* Common format for declarative workers, schedules, etc. This is only for client options.
+* File-based hierarchy. We don't need many levels of file merging for these options at this time.
+* Extreme focus on human authoring of this format. This is because it'd likely go against the "0 new dependencies" rule.
```

Comment:
I'd prioritize authoring DX / YAML over 0 deps

Reply:
It's really bad for libraries to force a dependency on all users of a library just for an optional side thing like this, that's really rough DX (we already suffer it so much with proto versions and that's an actual needed dependency).

Reply:
One of the scenarios Candace highlighted is for folks who are migrating production and dev environment code to cloud.  (We should align on whether that's a goal (P0/P1/P2)).

For such a use case, we might want to provide--and/or allow users to provide--extensions that allow configuration in whatever their system is, whether yaml or otherwise.
Also, if it's opt in as you say below, I don't see taking dependencies as a P0 problem.

Reply:
In any case, please reword this.  What "extreme focus on human authoring of this format" means is in the eye of the beholder.  Might be better to put something about human authoring features as a goal with a lower priority.  Then we can align on degrees of importance rather than stark goal/non-goal status.

Reply:
> Also, if it's opt in as you say below, I don't see taking dependencies as a P0 problem.

I think you're confusing opt-in from a caller POV and opt-in from a dependency POV. Someone can depend on our library and still have to opt-in to calling something on it, but just the act of depending on the library brings in dependencies. Dependencies are not opt-in, only the use of the call.

> Might be better to put something about human authoring features as a goal with a lower priority. Then we can align on degrees of importance rather than stark goal/non-goal status.

The ideal of the non-goal was to be a never priority. For this project there just happened to be no "future goals" like there are in other projects because meeting the goals is so simple as to not have any. Can I have priority infinite?